src/solver/impls/arkode/arkode_mri.cxx

```diff
@@ -0,0 +1,1090 @@
+ **************************************************************************/
+
+void ArkodeMRISolver::rhs_s(BoutReal t, BoutReal* udata, BoutReal* dudata) {
+  TRACE("Running RHS: ArkodeMRISolver::rhs_e({:e})", t);
+
+  // Load state from udata
+  load_vars(udata);
+
+  // Get the current timestep
+  // Note: ARKodeGetCurrentStep updated too late in older versions
+  ARKodeGetLastStep(arkode_mem, &hcur);
+
+  // Call RHS function
+  // run_rhs_s(t);
+  run_rhs_s(t);
+
+  // Save derivatives to dudata
+  save_derivs(dudata);
+}
+
+/**************************************************************************
+ * Fast RHS function du = F^f(t, u)
+ **************************************************************************/
+
+void ArkodeMRISolver::rhs_f(BoutReal t, BoutReal* udata, BoutReal* dudata) {
+  TRACE("Running RHS: ArkodeMRISolver::rhs_e({:e})", t);
+
+  // Load state from udata
+  load_vars(udata);
+
+  // Get the current timestep
+  // Note: ARKodeGetCurrentStep updated too late in older versions
+  ARKodeGetLastStep(arkode_mem, &hcur);
+
+  // Call RHS function
+  // run_rhs_f(t);
+  run_rhs_f(t);
+
+  // Save derivatives to dudata
+  save_derivs(dudata);
+}
+
+/**************************************************************************
+ * Preconditioner functions
+ **************************************************************************/
+
+void ArkodeMRISolver::pre_s(BoutReal t, BoutReal gamma, BoutReal delta, BoutReal* udata,
+                       BoutReal* rvec, BoutReal* zvec) {
+  TRACE("Running preconditioner: ArkodeMRISolver::pre({:e})", t);
+
+  const BoutReal tstart = bout::globals::mpi->MPI_Wtime();
+
+  if (!hasPreconditioner()) {
+    // Identity (but should never happen)
+    const auto length = N_VGetLocalLength_Parallel(uvec);
+    std::copy(rvec, rvec + length, zvec);
+    return;
+  }
+
+  // Load state from udata (as with res function)
+  load_vars(udata);
+
+  // Load vector to be inverted into F_vars
+  load_derivs(rvec);
+
+  runPreconditioner(t, gamma, delta);
+
+  // Save the solution from F_vars
+  save_derivs(zvec);
+
+  pre_Wtime_s += bout::globals::mpi->MPI_Wtime() - tstart;
+  pre_ncalls_s++;
+}
+
+void ArkodeMRISolver::pre_f(BoutReal t, BoutReal gamma, BoutReal delta, BoutReal* udata,
+                       BoutReal* rvec, BoutReal* zvec) {
+  TRACE("Running preconditioner: ArkodeMRISolver::pre({:e})", t);
+
+  const BoutReal tstart = bout::globals::mpi->MPI_Wtime();
+
+  if (!hasPreconditioner()) {
+    // Identity (but should never happen)
+    const auto length = N_VGetLocalLength_Parallel(uvec);
+    std::copy(rvec, rvec + length, zvec);
+    return;
+  }
+
+  // Load state from udata (as with res function)
+  load_vars(udata);
+
+  // Load vector to be inverted into F_vars
+  load_derivs(rvec);
+
+  runPreconditioner(t, gamma, delta);
+
+  // Save the solution from F_vars
+  save_derivs(zvec);
+
+  pre_Wtime_s += bout::globals::mpi->MPI_Wtime() - tstart;
+  pre_ncalls_s++;
```

Comment:
All of this needs to be updated, since the flags, functions, timers, and counters all correspond to the slow time scale.